UOP1_Project/Assets/Scripts/Sound/SoundEmitter.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Audio;
+
+[RequireComponent (typeof(AudioSource))]
+public class SoundEmitter : MonoBehaviour
+{
+	private AudioSource _audioSource;
+
+	public bool _initialPool = false;
+	public float _lastUseTimestamp = 0;
+
+	private void Awake()
+	{
+		DontDestroyOnLoad(this);
+		_audioSource = this.GetComponent<AudioSource>();
+	}
+
+	// Start is called before the first frame update
+	void Start()
+    {
+		_audioSource.playOnAwake = false;
+    }
+
+	public void PlaySound(AudioClip clip, SoundEmitterSettings settings, Vector3 position = default)
+	{
+		_audioSource.clip = clip;
+		ApplySettings(_audioSource, settings);
+		_audioSource.transform.position = position;
+		_lastUseTimestamp = Time.realtimeSinceStartup + clip.length;
```

Comment:
This will fail for looped sounds. Trying to think of a way to detect when IsInUse resolves to false for the first time that doesn't involve a coroutine polling for it.

Reply:
isPlaying is always true for looped sounds, the places where timestamp is checked we also check if isPlaying so for now no issues at all.

Reply:
> Using Mathf.infinity for loops would do the trick
 
Would this cause the looped source to never be trimmed even if a user called StopAllLoopedSounds?

Reply:
You are right, that's a problem. Do you think making the both the StopAllSounds and StopAllLoopedSound to just set the timestamp to te current timestamp to solve that?

Reply:
Yeah, I think that's a good solution.

Reply:
Hmm wondering if pooling and sound management are getting too coupled. Two thoughts:
1. Trimming is nice to have for memory but not a requisite for the object pool pattern. Maybe it's something that can be expanded upon?
2. I think it might be better to yield control of the SoundEmitter to the user. Then they could loop, play one shot, etc. to their heart's content. Once done, they can return it to the pool and you have a clear point at which the SoundEmitter is no longer in use.

Reply:
I don't understand first question, you say its not a requisite and its getting too coupled but want to expand on it?? I think the pooler is mostly done and I would not touch anything else for a first PR.

As for the second point, I programmed it centralized like this initially because that is what was asked in the forum thread, doing what you are purposing would be the complete opposite of what they are asking for.
From the thread:
> Characters, effects, events, can ask the audio system to play a sound ...

Reply:
Sorry. "Expand on" was a poor choice of words. I meant leave trimming functionality out for now and come back to it later.

And you're right, keeping control centralized fits the brief better. I was just thinking how easy it would be to Lerp position, volume, pitch, etc. if we had continuous control from user scripts.

Solid first PR I think. Nice work! 👍